lec10-modelling.Rmd

```diff
@@ -1,4 +1,707 @@
+```
+
+Let's plot just one of the habitats, sites, and treatments vs. time and colour by species.
+
+```{r}
+library(dplyr)
+library(tidyr)
+
+plant_data %>% 
+  filter(habitat == "Tundra") %>%
+  filter(site == 3) %>% 
+  filter(treatment == "rodentexclosure") %>% 
+  gather(Species, Value, -year, -treatment, -habitat, -site) %>% 
+  ggplot() +
+  geom_point(aes(x = year, y = Value, color = Species))
+```
+
+*Empetrum nigrum* looks like it could be following logistic growth, so let's try to fit 
+the data to that model. First, let's make a new variable with just the data we want to fit.
+
+```{r}
+e_nigrum <- plant_data %>% 
+  filter(site == 3) %>% 
+  filter(habitat == "Tundra") %>% 
+  filter(treatment == "rodentexclosure") %>% 
+  select(year, empetrum_nigrum) 
+
+```
+
+Now we define a logistic model function so that we can numerically generate the model's
+solution and compare it to the data.
+
+```{r}
+logistic_fn <- function(t, state, parameters) {
+  # Calculates dN/dt for the logistic equation
+  
+  # t: time point at which to evaluate derivative (doesn't actually change anything in this example)
+  # state: vector of variables (here it's just N)
+  # parameters: vector of model parameters c(r, K)
+  
+  N <- state 
+  
+  r <- parameters[1] # the first element of the parameters vector is r
+  K <- parameters[2] # the second element of the parameters vector is K
+
+  #rate of change
+  dN <- r * N * (1 - N / K)
+    
+  #return rate of change
+  return(list(c(dN)))
+}
+```
+
+Let's try running a single numerical solution and plotting it alongside the data.
+
+```{r}
+library(deSolve)
+
+parameters <- c(r = 0.5, K = 150)
+state <- c(N = e_nigrum[1,2]) # the first row and second column is the initial population size
+times <- seq(0, 14, by = 1) # make the same time vector as the data
+
+result <- ode(y = state, times = times, func = logistic_fn, parms = parameters)
+```
+
+```{r}
+# Plot
+result <- data.frame(result)
+
+p1 <- ggplot(e_nigrum) +
+  geom_point(aes(x = year, y = empetrum_nigrum))
+
+p1 +
+  geom_point(aes(x = time + 1998, y = N), result, color = 'blue') +
+  geom_line(aes(x = time + 1998, y = N), result, color = 'blue')
+```
+
+Now we'll define a grid of $r$ and $K$ values to try, then calculate a numerical solution 
+for each combination of parameters. 
+
+```{r}
+rvals = seq(0.05, 1.0, by = 0.05)
+Kvals = seq(120, 180, by = 5)
+
+# initialize empty matrix for likelihood calculation
+residuals_surface = matrix(, nrow = length(rvals), ncol = length(Kvals)) 
+
+# loop over values of r and k, calculate log likelihood
+for (i in seq(length(rvals))){
+  for (j in seq(length(Kvals))){
+    result <- ode(y = state, 
+                  times = times, 
+                  func = logistic_fn,
+                  parms = c(r = rvals[i], K = Kvals[j]))
+    result <- data.frame(result)
+    residuals <- result$N - e_nigrum$empetrum_nigrum
+    residuals_surface[i,j] <- sum(residuals^2)
+  }
+}
```

Comment:
There is a more 'R' way of doing this, but I don't have time to go into it. This is fine.